enzyme/Enzyme/MLIR/Analysis/DataFlowActivityAnalysis.cpp

```diff
@@ -0,0 +1,976 @@
+                  Value yieldOperand =
+                      linalgOp.getBlock()->getTerminator()->getOperand(
+                          resultIndex);
+                  auto *valueState =
+                      getOrCreate<BackwardValueActivity>(yieldOperand);
+                  propagateIfChanged(
+                      valueState,
+                      valueState->meet(ValueActivity::getActiveVal()));
+                }
+              }
+            }
+          }
+        });
+      }
+    }
+    propagateIfChanged(before, result);
+  }
+
+  void visitCallControlFlowTransfer(CallOpInterface call,
+                                    CallControlFlowAction action,
+                                    const BackwardMemoryActivity &after,
+                                    BackwardMemoryActivity *before) override {
+    meet(before, after);
+  }
+
+  void setToExitState(BackwardMemoryActivity *lattice) override {}
+
+private:
+  FunctionOpInterface parentOp;
+  ArrayRef<enzyme::Activity> argumentActivity;
+};
+
+void traverseCallGraph(FunctionOpInterface root,
+                       SymbolTableCollection *symbolTable,
+                       function_ref<void(FunctionOpInterface)> processFunc) {
+  std::deque<FunctionOpInterface> frontier{root};
+  DenseSet<FunctionOpInterface> visited{root};
+
+  while (!frontier.empty()) {
+    FunctionOpInterface curr = frontier.front();
+    frontier.pop_front();
+    processFunc(curr);
+
+    curr.walk([&](CallOpInterface call) {
+      auto neighbor = dyn_cast_if_present<FunctionOpInterface>(
+          call.resolveCallable(symbolTable));
+      if (neighbor && !visited.contains(neighbor)) {
+        frontier.push_back(neighbor);
+        visited.insert(neighbor);
+      }
+    });
+  }
+}
+
+void printActivityAnalysisResults(const DataFlowSolver &solver,
+                                  FunctionOpInterface callee,
+                                  const SmallPtrSet<Operation *, 2> &returnOps,
+                                  SymbolTableCollection *symbolTable,
+                                  bool verbose, bool annotate) {
+  auto isActiveData = [&](Value value) {
+    auto fva = solver.lookupState<ForwardValueActivity>(value);
+    auto bva = solver.lookupState<BackwardValueActivity>(value);
+    bool forwardActive = fva && fva->getValue().isActiveVal();
+    bool backwardActive = bva && bva->getValue().isActiveVal();
+    return forwardActive && backwardActive;
+  };
+
+  auto isConstantValue = [&](Value value) {
+    // TODO: integers/vectors that might be pointers
+    if (isa<LLVM::LLVMPointerType, MemRefType>(value.getType())) {
+      assert(returnOps.size() == 1);
+      auto *fma = solver.lookupState<ForwardMemoryActivity>(*returnOps.begin());
+      auto *bma = solver.lookupState<BackwardMemoryActivity>(
+          &callee.getFunctionBody().front().front());
+
+      auto *pointsToSets =
+          solver.lookupState<enzyme::PointsToSets>(*returnOps.begin());
```

Comment:
(*) why is this looking at points-to at the exit point of the function? Can't there be a situation where the value is known to _not yet_ point to some class and we are being excessively conservative here? Is there a case where that wouldn't be correct?

(Also, we'd need to handle the multi-return case correctly by joining the states at all return points).

Reply:
I think it's definitely a potential way to improve the precision of the results, but looking at intermediate steps requires some thought *at what point* the activity of a value is important. I'll think through this further.

Reply:
Ultimately, activity analysis is used to identify which operations are active. I suppose it is sufficient to reason about activity of values, which depends on points-to, _around_ that operation (before or after depending on the flow of activity) when deciding whether the operation is active.

Reply:
Concretely, I'm thinking about something like but with extra level of indirection

```mlir
%0 = ... // points to active value
store %0, %p  // %p transitively points to active value here
%a = load %p // %a points to active value (trivially)
// do something with %p
%1 = ... // points to constant values only
store %1, %p  // %p does *not* transitively point to active value here
%2 = load %p  // %2 points to constant values only trivially
// do something else with %p
```

if we are looking at the end of the function, and since we currently treat all stores as "may store" and join the points-to lattices, the return state will be that `%p` may be pointing to either constant or active pointer.